### PR TITLE
Enhanced authentication

### PR DESCRIPTION
**Pull request by tigercl**

### CODE REVIEW

File: src/emqx_protocol.erl [hunk 1]
```diff
@@ -277,36 +285,30 @@ process_packet(?CONNECT_PACKET(
...
+                    case authenticate(credentials(PState2), Password) of
+                        {ok, IsSuper} ->
+                            connack(open_session(PState2#pstate{is_super = IsSuper, auth_state = connected}));
+                        {error, Reason} ->
+                            ?LOG(error, "Username '~s' login failed for ~p", [Username, Reason], PState2),
+                            connack({?RC_NOT_AUTHORIZED, PState2})
+                    end
+            end;
+        {error, ReasonCode} ->
+            connack({ReasonCode, PState1})
```
Comment (spring2maz) [inline src/emqx_protocol.erl]: `connack/1` should now be changed to `connack/2`
i.e. use two args instead of a 2-elements tuple as a single arg.

Comment (spring2maz) [reply src/emqx_protocol.erl]: never mind, just notice line 303 is still the old style call.

File: src/emqx_protocol.erl [hunk 2]
```diff
@@ -566,6 +677,22 @@ send(Packet = ?PACKET(Type), PState = #pstate{proto_ver = Ver, sendfun = SendFun
...
+open_session(PState = #pstate{keepalive = Keepalive}) ->
+    PState1 = maybe_assign_client_id(PState),
+    case try_open_session(PState1) of
+        {ok, SPid, SP} ->
+            PState2 = PState1#pstate{session = SPid, connected = true},
+            ok = emqx_cm:register_connection(client_id(PState2), attrs(PState2)),
+            %% Start keepalive
+            start_keepalive(Keepalive, PState2),
+            %% Success
+            {?RC_SUCCESS, SP, PState2#pstate{auth_state = connected}};
```
Comment (spring2maz) [inline src/emqx_protocol.erl]: if we call `connack` here, it should be a continuation,
not a stack pop for the caller and continue there.

it's also easier to read if continue from here.

then we can convert `connack/1` to `connack/2`.

File: src/emqx_protocol.erl [hunk 3]
```diff
@@ -625,14 +752,20 @@ set_session_attrs({topic_alias_maximum, #pstate{zone = Zone, proto_ver = ProtoVe
...
     SessAttrs.
 
+authenticate(Credentials) ->
+    authenticate(Credentials, maps:get(password, Credentials, undefined)).
 
 authenticate(Credentials, Password) ->
     case emqx_access_control:authenticate(Credentials, Password) of
         ok -> {ok, false};
+        {ok, #{username := Username}} -> 
+            {ok, Username, false};
```
Comment (spring2maz) [inline src/emqx_protocol.erl]: need a macro from this `false`

File: src/emqx_protocol.erl [hunk 4]
```diff
@@ -441,6 +532,21 @@ connack({ReasonCode, PState = #pstate{proto_ver = ProtoVer}}) ->
...
+%%------------------------------------------------------------------------------
+%% Disconnect --> Client
+%%------------------------------------------------------------------------------
+
+disconnect({?RC_SUCCESS, PState}) ->
+    deliver({disconnect, ?RC_SUCCESS}, PState);
+disconnect({ReasonCode, PState = #pstate{proto_ver = ProtoVer}}) ->
+    ReasonCode1 = if ProtoVer =:= ?MQTT_PROTO_V5 ->
+                         ReasonCode;
+                     true ->
```
Comment (spring2maz) [inline src/emqx_protocol.erl]: align with `if`.
and a nitpicking: I personally find case clauses much more comfortable to read, so I never write `if`s

Comment (Gilbert-Wong) [reply src/emqx_protocol.erl]: So am i.

File: src/emqx_protocol.erl [hunk 5]
```diff
@@ -277,36 +285,30 @@ process_packet(?CONNECT_PACKET(
...
-          {error, ReasonCode} ->
-              {ReasonCode, PState1}
-      end);
+
+    case check_connect(Connect, PState1) of
+        {ok, PState2} ->
+            EnhancedAuth = is_map(ConnProps) andalso maps:is_key('Authentication-Method', ConnProps),
+            if 
+                EnhancedAuth ->
+                    process_enhanced_auth(maps:get('Authentication-Method', ConnProps, undefined),
```
Comment (spring2maz) [inline src/emqx_protocol.erl]: why not default to `<<>>` (replace `undefined`) the guard below checks `=:= undefined orelse =:= <<>>`.
and use a macro for it

Comment (spring2maz) [reply src/emqx_protocol.erl]: e.g. `?NO_AUTH_METHOD`?

File: src/emqx_protocol.erl (hunk 3 above)
Comment (spring2maz) [reply src/emqx_protocol.erl]: or use a variable name to hint what it is, e.g. `_IsSuper = false`

File: src/emqx_protocol.erl [hunk 6]
```diff
@@ -566,6 +677,22 @@ send(Packet = ?PACKET(Type), PState = #pstate{proto_ver = Ver, sendfun = SendFun
             {error, Reason}
     end.
 
+%% TODO: need a better function name
```
Comment (spring2maz) [inline src/emqx_protocol.erl]: name it as what it does: `ensure_client_id_and_try_open_session`

File: src/emqx_protocol.erl [hunk 7]
```diff
@@ -421,7 +423,96 @@ process_packet(?DISCONNECT_PACKET(?RC_SUCCESS, #{'Session-Expiry-Interval' := In
...
+    when AuthMethod =:= undefined orelse AuthMethod =:= <<>> ->
+    if 
+        AuthState =:= reauthenticating orelse AuthState =:= connected ->
+            disconnect({?RC_MALFORMED_PACKET, PState});
+        true ->
+            connack({?RC_MALFORMED_PACKET, PState})
+    end;
+process_enhanced_auth(AuthMethod, AuthData, PState = #pstate{auth_method = OldAuthMethod, auth_state = AuthState}) ->
+    if 
+        OldAuthMethod =/= <<>> andalso OldAuthMethod =/= AuthMethod ->
```
Comment (spring2maz) [inline src/emqx_protocol.erl]: this clause should be delegated to a function or variable with a more readable name.
e.g. `is_bad_auth_method(OldAuthMethod, AuthMethod)`

File: src/emqx_protocol.erl [hunk 8]
```diff
@@ -421,7 +423,96 @@ process_packet(?DISCONNECT_PACKET(?RC_SUCCESS, #{'Session-Expiry-Interval' := In
...
+            case authenticate(credentials(PState, #{auth_method => AuthMethod, 
+                                                    auth_data   => AuthData})) of
+                {continue, Reply} -> 
+                    deliver({auth, ?RC_CONTINUE_AUTHENTICATE, AuthMethod, Reply}, PState#pstate{auth_method = AuthMethod,
+                                                                                                auth_state = case AuthState of
+                                                                                                                 connected ->
+                                                                                                                     reauthenticating;
+                                                                                                                 _ ->
+                                                                                                                     authenticating
+                                                                                                             end});
```
Comment (Gilbert-Wong) [inline src/emqx_protocol.erl]: `case` and `end` should be aligned. And I suggest you to wrap your code. The code is toow long in a line to read.

File: src/emqx_protocol.erl [hunk 9]
```diff
@@ -102,7 +107,9 @@ init(#{peername := Peername, peercert := Peercert, sendfun := SendFun}, Options)
             enable_acl   = emqx_zone:get_env(Zone, enable_acl),
             recv_stats   = #{msg => 0, pkt => 0},
             send_stats   = #{msg => 0, pkt => 0},
-            connected    = false}.
+            connected    = false,
+            auth_method  = <<>>,
```
Comment (spring2maz) [inline src/emqx_protocol.erl]:
```suggestion
            auth_method  = ?NO_AUTH_METHOD,
```

File: src/emqx_protocol.erl [hunk 10]
```diff
@@ -421,7 +426,95 @@ process_packet(?DISCONNECT_PACKET(?RC_SUCCESS, #{'Session-Expiry-Interval' := In
 process_packet(?DISCONNECT_PACKET(?RC_SUCCESS), PState) ->
     {stop, normal, PState#pstate{will_msg = undefined}};
 process_packet(?DISCONNECT_PACKET(_), PState) ->
-    {stop, normal, PState}.
+    {stop, normal, PState};
+
+process_packet(?AUTH_PACKET(_), PState = #pstate{auth_state = disconnected}) ->
+    {error, proto_not_connected, PState};
+process_packet(?AUTH_PACKET(_), PState = #pstate{auth_method = <<>>, auth_state = AuthState}) ->
```
Comment (spring2maz) [inline src/emqx_protocol.erl]:
```suggestion
process_packet(?AUTH_PACKET(_), PState = #pstate{auth_method = ?NO_AUTH_METHOD, auth_state = AuthState}) ->
```

File: src/emqx_protocol.erl [hunk 11]
```diff
@@ -277,36 +288,30 @@ process_packet(?CONNECT_PACKET(
...
-              {ReasonCode, PState1}
-      end);
+
+    case check_connect(Connect, PState1) of
+        {ok, PState2} ->
+            EnhancedAuth = is_map(ConnProps) andalso maps:is_key('Authentication-Method', ConnProps),
+            if 
+                EnhancedAuth ->
+                    process_enhanced_auth(maps:get('Authentication-Method', ConnProps, ?NO_AUTH_METHOD), 
+                                          maps:get('Authentication-Data', ConnProps, ?NO_AUTH_METHOD),
```
Comment (spring2maz) [inline src/emqx_protocol.erl]: are you sure here to be `?NO_AUTH_METHOD` for `Autentication-Data` ?
maybe an other macro instead? e.g. `NO_AUTH_DATA`

File: src/emqx_protocol.erl [hunk 12]
```diff
@@ -421,7 +426,95 @@ process_packet(?DISCONNECT_PACKET(?RC_SUCCESS, #{'Session-Expiry-Interval' := In
...
+process_enhanced_auth(?NO_AUTH_METHOD, _AuthData, PState = #pstate{auth_state = AuthState}) ->
+    if 
+        AuthState =:= reauthenticating orelse AuthState =:= connected ->
+            disconnect({?RC_MALFORMED_PACKET, PState});
+        true ->
+            connack({?RC_MALFORMED_PACKET, PState})
+    end;
+process_enhanced_auth(AuthMethod, AuthData, PState = #pstate{auth_method = OldAuthMethod, auth_state = AuthState}) ->
+    if 
+        OldAuthMethod =/= <<>> andalso OldAuthMethod =/= AuthMethod ->
```
Comment (spring2maz) [inline src/emqx_protocol.erl]:
```suggestion
        case OldAuthMethod =/= ?NO_AUTH_METHOD andalso OldAuthMethod =/= AuthMethod of
            true when AuthState =:= reauthenticating orelse AuthState =:= connected ->
                ...
            true ->
                ...
             false ->
                ...
        end
```

File: src/emqx_protocol.erl (hunk 11 above)
Comment (tigercl) [reply src/emqx_protocol.erl]: I am sorry for this

File: src/emqx_protocol.erl (hunk 2 above)
Comment (tigercl) [reply src/emqx_protocol.erl]: {?RC_SUCCESS, SP, PState} -> connack({?RC_SUCCESS, SP, PState})?

File: src/emqx_protocol.erl [hunk 13]
```diff
@@ -421,7 +427,95 @@ process_packet(?DISCONNECT_PACKET(?RC_SUCCESS, #{'Session-Expiry-Interval' := In
...
+    if 
+        AuthState =:= authenticating ->
+            connack({?RC_MALFORMED_PACKET, PState});
+        true ->     % reauthenticating or connected 
+            disconnect({?RC_MALFORMED_PACKET, PState})
+    end;
+process_packet(?AUTH_PACKET(?RC_CONTINUE_AUTHENTICATE, _), PState = #pstate{auth_state = connected}) -> 
+    disconnect({?RC_PROTOCOL_ERROR, PState});
+process_packet(?AUTH_PACKET(?RC_CONTINUE_AUTHENTICATE, Properties), PState) ->
+    process_enhanced_auth(maps:get('Authentication-Method', Properties, ?NO_AUTH_METHOD), maps:get('Authentication-Data', Properties, ?NO_AUTH_DATA), PState);
```
Comment (terry-xiaoyu) [inline src/emqx_protocol.erl]: Change `maps:get('Authentication-Method', Properties, ?NO_AUTH_METHOD)` to `auth_method(Properties)`

File: src/emqx_protocol.erl [hunk 14]
```diff
@@ -421,7 +427,95 @@ process_packet(?DISCONNECT_PACKET(?RC_SUCCESS, #{'Session-Expiry-Interval' := In
...
+process_packet(?AUTH_PACKET(?RC_RE_AUTHENTICATE, Properties), PState = #pstate{auth_state = connected}) ->
+    process_enhanced_auth(maps:get('Authentication-Method', Properties, ?NO_AUTH_METHOD), maps:get('Authentication-Data', Properties, ?NO_AUTH_DATA), PState);
```
Comment (terry-xiaoyu) [inline src/emqx_protocol.erl]: Change `maps:get('Authentication-Data', Properties, ?NO_AUTH_DATA)` to `auth_data(Properties)`

File: src/emqx_protocol.erl [hunk 15]
```diff
@@ -421,7 +427,95 @@ process_packet(?DISCONNECT_PACKET(?RC_SUCCESS, #{'Session-Expiry-Interval' := In
...
+process_enhanced_auth(?NO_AUTH_METHOD, _AuthData, PState = #pstate{auth_state = AuthState}) ->
+    if 
+        AuthState =:= reauthenticating orelse AuthState =:= connected ->
+            disconnect({?RC_MALFORMED_PACKET, PState});
+        true ->
+            connack({?RC_MALFORMED_PACKET, PState})
+    end;
+process_enhanced_auth(AuthMethod, AuthData, PState = #pstate{auth_method = OldAuthMethod, auth_state = AuthState}) ->
+    if 
+        OldAuthMethod =/= ?NO_AUTH_METHOD andalso OldAuthMethod =/= AuthMethod ->
```
Comment (terry-xiaoyu) [inline src/emqx_protocol.erl]: I think this kind of `if` clauses could be optimized by function clauses.

**Pull request closed without merging**